Index: libreswan-3.18/programs/pluto/ikev1_xauth.c
===================================================================
--- libreswan-3.18.orig/programs/pluto/ikev1_xauth.c
+++ libreswan-3.18/programs/pluto/ikev1_xauth.c
@@ -1471,6 +1471,7 @@ static void log_bad_attr(const char *kin
  * @param md Message Digest
  * @return stf_status
  */
+extern size_t strnlen(const char *s, size_t maxlen); /* lino: avoid implicit declaration of function strnleng */
 stf_status xauth_inR0(struct msg_digest *md)
 {
 	pb_stream *attrs = &md->chain[ISAKMP_NEXT_MCFG_ATTR]->pbs;
