Index: openswan-2.6.49/linux/net/ipsec/ipsec_tunnel.c
===================================================================
--- openswan-2.6.49.orig/linux/net/ipsec/ipsec_tunnel.c
+++ openswan-2.6.49/linux/net/ipsec/ipsec_tunnel.c
@@ -314,6 +314,7 @@ int klips_header_parse(const struct sk_b
 	return ret;
 }
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 DEBUG_NO_STATIC int
 klips_rebuild_header(struct sk_buff *skb)
 {
@@ -375,6 +376,7 @@ klips_rebuild_header(struct sk_buff *skb
 	skb->dev = tmp;
 	return ret;
 }
+#endif
 
 int klips_header_cache(
 	const struct neighbour *neigh,
@@ -478,7 +480,9 @@ klips_header_cache_update(struct hh_cach
 const struct header_ops klips_header_ops ____cacheline_aligned = {
 	.create		= klips_header,
 	.parse		= klips_header_parse,
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 	.rebuild	= klips_rebuild_header,
+#endif
 	.cache		= klips_header_cache,
 	.cache_update	= klips_header_cache_update,
 };
@@ -1165,6 +1169,7 @@ ipsec_tunnel_hard_header(struct sk_buff
 	return ret;
 }
 
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 DEBUG_NO_STATIC int
 ipsec_tunnel_rebuild_header(struct sk_buff *skb)
 {
@@ -1236,6 +1241,7 @@ ipsec_tunnel_rebuild_header(struct sk_bu
 	skb->dev = tmp;
 	return ret;
 }
+#endif
 
 #ifdef HAVE_SET_MAC_ADDR
 DEBUG_NO_STATIC int
@@ -1349,7 +1355,9 @@ ipsec_tunnel_cache_update(struct hh_cach
 #ifdef HAVE_NETDEV_HEADER_OPS
 const struct header_ops ipsec_tunnel_header_ops = {
 	.create		= ipsec_tunnel_hard_header,
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 	.rebuild	= ipsec_tunnel_rebuild_header,
+#endif
 	.cache_update	= ipsec_tunnel_cache_update,
 };
 #endif
@@ -1430,7 +1438,9 @@ ipsec_tunnel_detach(struct net_device *d
 	prv->header_ops = NULL;
 #else
 	prv->hard_header = NULL;
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 	prv->rebuild_header = NULL;
+#endif
 	prv->header_cache_update = NULL;
 #endif
 	prv->set_mac_address = NULL;
@@ -1441,7 +1451,9 @@ ipsec_tunnel_detach(struct net_device *d
 	dev->header_ops = NULL;
 #else
 	dev->hard_header = NULL;
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 	dev->rebuild_header = NULL;
+#endif
 	dev->header_cache_update = NULL;
 	dev->neigh_setup        = NULL;
 #endif
@@ -2005,7 +2017,9 @@ ipsec_tunnel_init(struct net_device *dev
 	dev->header_ops		= NULL;
 #else
 	dev->hard_header	= NULL;
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0))
 	dev->rebuild_header 	= NULL;
+#endif
 	dev->header_cache_update= NULL;
 #endif
 #ifdef HAVE_NET_DEVICE_OPS
